File: script-files/measures/outputsContinuous/f.js
jStat = require("../node_modules/jstat");

dists = {
    f: {
        measurements: {
            nonCentralMoments: function (k, df1, df2) {
                return ((df2 / df1) ** k) * (jStat.gammafn(df1 / 2 + k) / jStat.gammafn(df1 / 2)) * (jStat.gammafn(df2 / 2 - k) / jStat.gammafn(df2 / 2))
            },
            centralMoments: function (k, df1, df2) {
                const miu1 = this.nonCentralMoments(1, df1, df2);
                const miu2 = this.nonCentralMoments(2, df1, df2);
                const miu3 = this.nonCentralMoments(3, df1, df2);
                const miu4 = this.nonCentralMoments(4, df1, df2);

                let result;
                switch (k) {
                    case 1: result = 0; break;
                    case 2: result = miu2 - miu1 ** 2; break;
                    case 3: result = miu3 - 3 * miu1 * miu2 + 2 * miu1 ** 3; break;
                    case 4: result = miu4 - 4 * miu1 * miu3 + 6 * (miu1 ** 2) * miu2 - 3 * (miu1 ** 4); break;
                };
                return result
            },
            stats: {
                mean: function (df1, df2) {
                    const miu1 = dists.f.measurements.nonCentralMoments(1, df1, df2);
                    return miu1;
                },
                variance: function (df1, df2) {
                    const miu1 = dists.f.measurements.nonCentralMoments(1, df1, df2);
                    const miu2 = dists.f.measurements.nonCentralMoments(2, df1, df2);
                    return miu2 - miu1 ** 2;
                },
                standardDeviation: function (df1, df2) {
                    return Math.sqrt(this.variance(df1, df2));
                },
                skewness: function (df1, df2) {
                    const central_miu3 = dists.f.measurements.centralMoments(3, df1, df2);
                    return central_miu3 / (this.standardDeviation(df1, df2) ** 3);
                },
                kurtosis: function (df1, df2) {
                    const central_miu4 = dists.f.measurements.centralMoments(4, df1, df2);
                    return central_miu4 / (this.standardDeviation(df1, df2) ** 4);
                },
                median: function (df1, df2) {
                    return dists.f.measurements.ppf(0.5, df1, df2);
                },
                mode: function (df1, df2) {
                    return (df2 * (df1 - 2)) / (df1 * (df2 + 2));
                },
            },
        }
    }
}
console.log(dists.f.measurements.stats.mean(15, 20))
console.log(dists.f.measurements.stats.variance(15, 20))
console.log(dists.f.measurements.stats.standardDeviation(15, 20))
console.log(dists.f.measurements.stats.skewness(15, 20))
console.log(dists.f.measurements.stats.kurtosis(15, 20))
// console.log(dists.f.measurements.stats.median(15, 20))
console.log(dists.f.measurements.stats.mode(15, 20))

// mean_value: 1.1111111111111112
// variance_value: 0.3395061728395059
// standard_deviation_value: 0.5826715823167506
// skewness_value: 1.7434744489062268
// kurtosis_value: 9.319480519480532
// median_value: 0.9886983373954129
// mode_value: 0.7878787878787878

File: phitter/continuous/continuous_distributions/f.py
import numpy
import scipy.special
import scipy.stats


class F:
    """
    F distribution
    - Parameters F Distribution: {"df1": *, "df2": *}
    - https://phitter.io/distributions/continuous/f
    """

    def __init__(
        self,
        parameters: dict[str, int | float] = None,
        continuous_measures=None,
        init_parameters_examples=False,
    ):
        """
        - Initializes the F Distribution by either providing a Continuous Measures instance [ContinuousMeasures] or a dictionary with the distribution's parameters.
        - Parameters F Distribution: {"df1": *, "df2": *}
        - https://phitter.io/distributions/continuous/f
        """
        if continuous_measures is None and parameters is None and init_parameters_examples == False:
            raise ValueError(
                "You must initialize the distribution by providing one of the following: distribution parameters, a Continuous Measures [ContinuousMeasures] instance, or by setting init_parameters_examples to True."
            )
        if continuous_measures != None:
            self.parameters = self.get_parameters(continuous_measures=continuous_measures)
        if parameters != None:
            self.parameters = parameters
        if init_parameters_examples:
            self.parameters = self.parameters_example

        self.df1 = self.parameters["df1"]
        self.df2 = self.parameters["df2"]

    @property
    def name(self):
        return "f"

    @property
    def parameters_example(self) -> dict[str, int | float]:
        return {"df1": 5, "df2": 5}

    def cdf(self, x: float | numpy.ndarray) -> float | numpy.ndarray:
        """
        Cumulative distribution function
        """
        # result = scipy.special.betainc(self.df1 / 2, self.df2 / 2, x * self.df1 / (self.df1 * x + self.df2))
        result = scipy.stats.f.cdf(x, self.df1, self.df2)
        return result

    def pdf(self, x: float | numpy.ndarray) -> float | numpy.ndarray:
        """
        Probability density function
        """
        # result = (1 / scipy.special.beta(self.df1 / 2, self.df2 / 2)) * ((self.df1 / self.df2) ** (self.df1 / 2)) * (x ** (self.df1 / 2 - 1)) * ((1 + x * self.df1 / self.df2) ** (-1 * (self.df1 + self.df2) / 2))
        result = scipy.stats.f.pdf(x, self.df1, self.df2)
        return result

    def ppf(self, u: float | numpy.ndarray) -> float | numpy.ndarray:
        """
        Percent point function. Inverse of Cumulative distribution function. If CDF[x] = u => PPF[u] = x
        """
        t = scipy.special.betaincinv(self.df1 / 2, self.df2 / 2, u)
        result = (self.df2 * t) / (self.df1 * (1 - t))
        return result

    def sample(self, n: int, seed: int | None = None) -> numpy.ndarray:
        """
        Sample of n elements of ditribution
        """
        if seed:
            numpy.random.seed(seed)
        return self.ppf(numpy.random.rand(n))

    def non_central_moments(self, k: int) -> float | None:
        """
        Parametric no central moments. µ[k] = E[Xᵏ] = ∫xᵏ∙f(x) dx
        """
        return (self.df2 / self.df1) ** k * (scipy.special.gamma(self.df1 / 2 + k) / scipy.special.gamma(self.df1 / 2)) * (scipy.special.gamma(self.df2 / 2 - k) / scipy.special.gamma(self.df2 / 2))

    def central_moments(self, k: int) -> float | None:
        """
        Parametric central moments. µ'[k] = E[(X - E[X])ᵏ] = ∫(x-µ[k])ᵏ∙f(x) dx
        """
        µ1 = self.non_central_moments(1)
        µ2 = self.non_central_moments(2)
        µ3 = self.non_central_moments(3)
        µ4 = self.non_central_moments(4)

        if k == 1:
            return 0
        if k == 2:
            return µ2 - µ1**2
        if k == 3:
            return µ3 - 3 * µ1 * µ2 + 2 * µ1**3
        if k == 4:
            return µ4 - 4 * µ1 * µ3 + 6 * µ1**2 * µ2 - 3 * µ1**4

        return None

    @property
    def mean(self) -> float:
        """
        Parametric mean
        """
        µ1 = self.non_central_moments(1)
        return µ1

    @property
    def variance(self) -> float:
        """
        Parametric variance
        """
        µ1 = self.non_central_moments(1)
        µ2 = self.non_central_moments(2)
        return µ2 - µ1**2

    @property
    def standard_deviation(self) -> float:
        """
        Parametric standard deviation
        """
        return numpy.sqrt(self.variance)

    @property
    def skewness(self) -> float:
        """
        Parametric skewness
        """
        central_µ3 = self.central_moments(3)
        return central_µ3 / self.standard_deviation**3

    @property
    def kurtosis(self) -> float:
        """
        Parametric kurtosis
        """
        central_µ4 = self.central_moments(4)
        return central_µ4 / self.standard_deviation**4

    @property
    def median(self) -> float:
        """
        Parametric median
        """
        return self.ppf(0.5)

    @property
    def mode(self) -> float:
        """
        Parametric mode
        """
        return (self.df2 * (self.df1 - 2)) / (self.df1 * (self.df2 + 2))

    @property
    def num_parameters(self) -> int:
        """
        Number of parameters of the distribution
        """
        return len(self.parameters)

    def parameter_restrictions(self) -> bool:
        """
        Check parameters restrictions
        """
        v1 = self.df1 > 0
        v2 = self.df2 > 0
        return v1 and v2

    def get_parameters(self, continuous_measures) -> dict[str, float | int]:
        """
        Calculate proper parameters of the distribution from sample continuous_measures.
        The parameters are calculated by formula.

        Parameters
        ==========
        continuous_measures: MEASUREMESTS
            attributes: mean, std, variance, skewness, kurtosis, median, mode, min, max, size, num_bins, data

        Returns
        =======
        parameters: {"df1": *, "df2": *}
        """
        ## Scipy parameters of distribution
        scipy_parameters = scipy.stats.f.fit(continuous_measures.data_to_fit)

        ## Results
        parameters = {"df1": scipy_parameters[0], "df2": scipy_parameters[1]}

        return parameters


if __name__ == "__main__":
    import sys

    sys.path.append("../")
    from continuous_measures import ContinuousMeasures

    ## Import function to get continuous_measures
    def get_data(path: str) -> list[float]:
        sample_distribution_file = open(path, "r")
        data = [float(x.replace(",", ".")) for x in sample_distribution_file.read().splitlines()]
        sample_distribution_file.close()
        return data

    ## Distribution class
    path = "../../../distributions_samples/continuous_distributions_sample/sample_f.txt"
    data = get_data(path)
    continuous_measures = ContinuousMeasures(data)
    distribution = F(continuous_measures=continuous_measures)

    print(f"{distribution.name} distribution")
    print(f"Parameters: {distribution.parameters}")
    print(f"CDF: {distribution.cdf(continuous_measures.mean)} {distribution.cdf(numpy.array([continuous_measures.mean, continuous_measures.mean]))}")
    print(f"PDF: {distribution.pdf(continuous_measures.mean)} {distribution.pdf(numpy.array([continuous_measures.mean, continuous_measures.mean]))}")
    print(f"PPF: {distribution.ppf(0.5)} {distribution.ppf(numpy.array([0.5, 0.5]))} - V: {distribution.cdf(distribution.ppf(0.5))}")
    print(f"SAMPLE: {distribution.sample(5)}")
    print(f"\nSTATS")
    print(f"mean: {distribution.mean} - {continuous_measures.mean}")
    print(f"variance: {distribution.variance} - {continuous_measures.variance}")
    print(f"skewness: {distribution.skewness} - {continuous_measures.skewness}")
    print(f"kurtosis: {distribution.kurtosis} - {continuous_measures.kurtosis}")
    print(f"median: {distribution.median} - {continuous_measures.median}")
    print(f"mode: {distribution.mode} - {continuous_measures.mode}")
